File: app/src/main/java/sjj/novel/view/reader/page/TxtLine.java
package sjj.novel.view.reader.page;

import java.util.Objects;

public class TxtLine {
    public CharSequence txt;
    public boolean isTitle;


    public int top;
    public int bottom;
    public int left;
    public int right;

    /**
     * 当前行号
     */
    public int index;
    public int charStart;

    public TxtLine(CharSequence txt,boolean isTitle) {
        this.txt = txt;
        this.isTitle = isTitle;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TxtLine line = (TxtLine) o;
        return index == line.index && Float.compare(line.top, top) == 0 && Float.compare(line.bottom, bottom) == 0 && Float.compare(line.left, left) == 0 && Float.compare(line.right, right) == 0 && Objects.equals(txt, line.txt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(txt, index, top, bottom, left, right);
    }

    @Override
    public String toString() {
        return "TxtLine{" +
                "txt='" + txt + '\'' +
                ", isTitle=" + isTitle +
                ", top=" + top +
                ", bottom=" + bottom +
                ", left=" + left +
                ", right=" + right +
                ", index=" + index +
                ", charStart=" + charStart +
                '}';
    }
}

File: app/src/main/java/sjj/novel/view/reader/page/TxtLine.kt
package sjj.novel.view.reader.page

class TxtLine(
    var txt: String,
    var isTitle: Boolean,
    var height: Float,
    var width: Float,
    var isParaEnd: Boolean
) {

    var top = 0f


    var bottom = 0f


    var left = 0f


    var right = 0f

    /**
     * 当前行号
     */

    var index = 0


    var clusterStart = 0


    var clusterLeft = FloatArray(0)


    var clusterRight = FloatArray(0)

    /** grapheme cluster boundaries: [0, end1, end2, ..., txt.length] */

    var clusterBoundaries: IntArray = IntArray(0)

    fun setLeftAndRight(left: FloatArray, right: FloatArray, size: Int) {
        this.clusterLeft = left.copyOf(size)
        this.clusterRight = right.copyOf(size)
    }

    override fun toString(): String {
        return "TxtLine{" +
                "txt='" + txt + '\'' +
                ", isTitle=" + isTitle +
                ", top=" + top +
                ", bottom=" + bottom +
                ", left=" + left +
                ", right=" + right +
                ", index=" + index +
                ", clusterStart=" + clusterStart +
                '}'
    }


}